### PR TITLE
DBTP-603 Reintroduce basic auth

### PR DESCRIPTION
**Pull request by gabelton**
Basic auth logic should be unchanged from original ip-filter project. Tried to simplify where possible, adding a few comments and docstrings.

I've lifted the tests directly from [here](https://github.com/uktrade/dit-cf-security/blob/master/tests.py#L1614C4-L1614C4).

Also added some pre-commit checks, using the same style conventions as in copilot-tools repo.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+          ]
+  - repo: https://github.com/pycqa/isort
+    rev: 5.12.0
+    hooks:
+      - id: isort
+        args: [--force-single-line-imports, --profile, black]
+  - repo: https://github.com/ikamensh/flynt/
+    rev: '1.0.0'
+    hooks:
+      - id: flynt
```
Comment (acodeninja) [inline .pre-commit-config.yaml:32]: No new line on EOF

File: tests.py [hunk 2]
```diff
@@ -1408,19 +1399,19 @@ def test_x_forwarded_for_index_respected(self):
...
-                url="http://127.0.0.1:8080/",
-                headers={
-                    "host": "somehost.com",
-                    "x-forwarded-for": "1.2.3.4, 1.1.1.1",
-                },
-            )
-            .status
-        )
-        self.assertEqual(status, 200)
+        # status = (
```
Comment (acodeninja) [inline tests.py:1402]: Can we just delete instead of commenting out?

File: tests.py [hunk 3]
```diff
@@ -1570,6 +1563,547 @@ def test_can_process_request_with_ipv6_ip(self):
...
+                ),
+            )
+        )
+
+        self.addCleanup(create_origin(8081))
+        wait_until_connectable(8080)
+        wait_until_connectable(8081)
+        wait_until_connectable(2772)
+
+        status = (
```
Comment (acodeninja) [inline tests.py:1713]: A lot of these calls are pretty similar, maybe a test helper method that accepts the bits that vary will make these tests more readable?

File: utils.py [hunk 4]
```diff
@@ -0,0 +1,8 @@
+def constant_time_is_equal(a, b):
```
Comment (acodeninja) [inline utils.py:1]: This looks the same as the one in `auth.py` can we remove one and keep the other. Seems like the `utils.py` one is the one to keep.

File: requirements.in [hunk 5]
```diff
@@ -2,6 +2,6 @@ flask==2.3.2
 gunicorn
 urllib3
 jinja2
-
+pre-commit
```
Comment (yusufsheiqh) [inline requirements.in:5]: We should move this over to the `requirements-test.in` file and use the `requirements-test.txt` file.

**Review by yusufsheiqh — APPROVED** [review]

**Pull request merged**